instrumentation-core/spring/README.md

```diff
@@ -340,6 +341,323 @@ After running Jaeger locally, navigate to the url below. Make sure to refresh th
+
+   public static void main(String[] args) throws IOException {
+      SpringApplication.run(SecondServiceApplication.class, args);
+   }
+}
+```
+
+Add the RestController below to your SecondService project. This controller will return a string when SecondServiceController.secondTracedMethod is called:
+
+```java
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+
+@RestController
+@RequestMapping(value = "/time")
+public class SecondServiceController {
+   @Autowired
+   private Tracer tracer;
+
+   @GetMapping
+   public String secondTracedMethod() {
+      return "It's time to get a watch";
+   }
+}
+```
+
+#### Create Controller Interceptor
+
+Add the class below to wrap all requests to the SecondServiceController in a span. This class will call the preHandle method before the RestController is entered and the postHandle method after a response is created. 
+
+The preHandle method starts a span for each request. The postHandle method closes the span and adds the span context to the response header. This implementation is shown below:    
+
+```java
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+import io.grpc.Context;
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.TracingContextUtils;
+
+@Component
+public class ControllerTraceInterceptor implements HandlerInterceptor {
+
+   private static final HttpTextFormat.Getter<HttpServletRequest> getter =
+         new HttpTextFormat.Getter<HttpServletRequest>() {
+            public String get(HttpServletRequest req, String key) {
+               return req.getHeader(key);
+            }
+         };
+
+   private static final HttpTextFormat.Setter<HttpServletResponse> setter =
+         new HttpTextFormat.Setter<HttpServletResponse>() {
+            @Override
+            public void set(HttpServletResponse response, String key, String value) {
+               response.addHeader(key, value);
+            }
+         };
+
+   @Autowired
+   private Tracer tracer;
+
+   @Override
+   public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+         throws Exception {
+     
+      Span span = createSpanWithParent(request, context);
+      span.setAttribute("handler", "pre");
+      tracer.withSpan(span);
```

Comment:
it looks like this `Scope` is never closed?

Reply:
this is important to address because the scope is bound to `ThreadLocal` and so will leak, and if/when that thread is reused it will have wrong/old span bound to it

Reply:
Interesting. I noticed that the HandlerInterceptor.postHandle is not called when an unhandled exception is thrown. So I moved this code to afterCompletion but that doesn't address this threading issue. Would you recommend implementing a Serverlet Filter instead of a HandlerInterceptor to wrap controllers in a span?

Reply:
servlet filter sounds good 👍